torch/optim/adamw.py

```diff
@@ -371,8 +404,22 @@ def _single_tensor_adamw(
         # Perform stepweight decay
         param.mul_(1 - lr * weight_decay)
 
+        device = param.device
+
+        device = param.device
+        dtype = param.dtype
+
+        if beta1_dict is not None:
+            key = (device, dtype)
+            if key not in beta1_dict:
+                beta1_dict[key] = beta1.to(device=device, dtype=dtype, non_blocking=True)  # type: ignore[union-attr]
```

Comment:
Why dtype as well? Is there a case where beta1 dtype shouldn't stay what the user input'd?

Reply:
I needed dtype here for double. Specifically an op below didn't support upcasting the beta (iirc it was addcdiv). By default, the single elem tensor will be float32. I think the foreach ops upcast so this wasn't needed in the multitensor case.

Reply:
Hm in this case, the beta would downcast if the params were bf16 or smaller though, which is not desirable. Moreover, the scalar tensor should not run into casting issues for single tensor lerp_ (lerp should do type promotion correctly).

Reply:
Do you have the error for when you don't have the dtype?

Reply:
Yeah let me find the error and we can discuss, if it's actually a type promotion bug I can just file an issue. Currently on a plane tho so I'll do it when I land 😅

Reply:
Yeah it still happens - https://gist.github.com/mlazos/1ec56b8a09682105f9b4c17990aed78d

When I remove the dtype from the key it fails with float64 when dtype of the beta is float.

Reply:
yea, upcasting only when dtype is fp64 + filing an issue sounds good to me!